### ISSUE TITLE
mount: Folder is still uploaded if renamed and deleted before --vfs-write-back time expired.

### ISSUE THREAD

**Issue by jdr85** [rclone/rclone#6809]
#### The associated forum post URL from `https://forum.rclone.org`
https://forum.rclone.org/t/borg-lock-exclusive-recreated-after-vfs-write-back/36524

#### What is the problem you are having with rclone?
If I mount a cloud storage with --vfs-cache-mode=writes or full, rclone does not seem to handle folder renames and deletions properly. Steps to reproduce:

Mount cloud to /mnt/tmp:
`
rclone -vv mount --cache-dir=/tmp/rclone_vfscache/ --vfs-cache-mode=writes --vfs-cache-max-age=2m --vfs-write-back=60s --vfs-cache-max-size 1M pCloud: /mnt/tmp
`

The following steps must be done within the vfs-write-back time (which can of course be set to something longer than 60 s).
Create folder old in /mnt/tmp:
`
mkdir old
`

Create file in old:
`
touch old/file
`

rename old to new:
`
mv old new
`

delete new:
`
rm -rf new
`

Result:
* At first, everything looks OK both in /mnt/tmp and cloud WebGUI. Nothing is present.
* But after the vfs-write-back time has expired, new/file is uploaded to the cloud and visible in the WebGUI.
* After the dir-cache-time (I guess) has expired, it is also visible in /mnt/tmp

If the folder is deleted AFTER the vfs-write-back-time has expired, everything works as expected.

#### What is your rclone version (output from `rclone version`)
rclone v1.61.1
- os/version: debian 11.6 (64 bit)
- os/kernel: 5.10.0-21-amd64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.19.4
- go/linking: static
- go/tags: none

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
Debian 11.6 x86_64

#### Which cloud storage system are you using? (e.g. Google Drive)
Seems to be independent of cloud storage.
* pCloud
* GMX via WebDAV
* Getspace via WebDAV

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
`rclone -vv mount --cache-dir=/tmp/rclone_vfscache/ --vfs-cache-mode=writes --vfs-cache-max-age=2m --vfs-write-back=60s --vfs-cache-max-size 1M pCloud: /mnt/tmp`

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)
[rclone.log](https://github.com/rclone/rclone/files/10895714/rclone.log)
```
2023/03/06 08:18:42 DEBUG : rclone: Version "v1.61.1" starting with parameters ["rclone" "-vv" "--log-file=/tmp/rclone.log" "mount" "--cache-dir=/tmp/rclone_vfscache/" "--vfs-cache-mode=writes" "--vfs-cache-max-age=2m" "--vfs-write-back=60s" "--vfs-cache-max-size" "1M" "pCloud:" "/mnt/tmp"]
2023/03/06 08:18:42 DEBUG : Creating backend with remote "pCloud:"
2023/03/06 08:18:42 DEBUG : Using config file from "/daten/_jdr/rclone.conf"
2023/03/06 08:18:42 INFO  : pcloud root '': poll-interval is not supported by this remote
2023/03/06 08:18:42 DEBUG : vfs cache: root is "/tmp/rclone_vfscache"
2023/03/06 08:18:42 DEBUG : vfs cache: data root is "/tmp/rclone_vfscache/vfs/pCloud"
2023/03/06 08:18:42 DEBUG : vfs cache: metadata root is "/tmp/rclone_vfscache/vfsMeta/pCloud"
2023/03/06 08:18:42 DEBUG : Creating backend with remote "/tmp/rclone_vfscache/vfs/pCloud/"
2023/03/06 08:18:42 DEBUG : fs cache: renaming cache item "/tmp/rclone_vfscache/vfs/pCloud/" to be canonical "/tmp/rclone_vfscache/vfs/pCloud"
2023/03/06 08:18:42 DEBUG : Creating backend with remote "/tmp/rclone_vfscache/vfsMeta/pCloud/"
2023/03/06 08:18:42 DEBUG : fs cache: renaming cache item "/tmp/rclone_vfscache/vfsMeta/pCloud/" to be canonical "/tmp/rclone_vfscache/vfsMeta/pCloud"
2023/03/06 08:18:42 DEBUG : pcloud root '': Mounting on "/mnt/tmp"
2023/03/06 08:18:42 INFO  : vfs cache: cleaned: objects 0 (was 0) in use 0, to upload 0, uploading 0, total size 0 (was 0)
2023/03/06 08:18:42 DEBUG : : Root: 
2023/03/06 08:18:42 DEBUG : : >Root: node=/, err=<nil>
2023/03/06 08:18:49 DEBUG : /: Attr: 
2023/03/06 08:18:49 DEBUG : /: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2023/03/06 08:19:38 DEBUG : /: Attr: 
2023/03/06 08:19:38 DEBUG : /: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2023/03/06 08:19:38 DEBUG : /: Lookup: name="old"
2023/03/06 08:19:38 DEBUG : /: >Lookup: node=<nil>, err=no such file or directory
2023/03/06 08:19:38 DEBUG : /: Mkdir: name="old"
2023/03/06 08:19:39 DEBUG : : Added virtual directory entry vAddDir: "old"
2023/03/06 08:19:39 DEBUG : /: >Mkdir: node=old/, err=<nil>
2023/03/06 08:19:39 DEBUG : old/: Attr: 
2023/03/06 08:19:39 DEBUG : old/: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2023/03/06 08:19:42 INFO  : vfs cache: cleaned: objects 0 (was 0) in use 0, to upload 0, uploading 0, total size 0 (was 0)
2023/03/06 08:19:51 DEBUG : /: Attr: 
2023/03/06 08:19:51 DEBUG : /: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2023/03/06 08:19:51 DEBUG : old/: Lookup: name="file"
2023/03/06 08:19:51 DEBUG : old/: >Lookup: node=<nil>, err=no such file or directory
2023/03/06 08:19:51 DEBUG : old/: Create: name="file"
2023/03/06 08:19:51 DEBUG : old/file: Open: flags=O_WRONLY|O_CREATE|0x800
2023/03/06 08:19:51 DEBUG : old/file: newRWFileHandle: 
2023/03/06 08:19:51 DEBUG : old/file(0xc0005781c0): openPending: 
2023/03/06 08:19:51 DEBUG : old/file: vfs cache: truncate to size=0 (not needed as size correct)
2023/03/06 08:19:51 DEBUG : old: Added virtual directory entry vAddFile: "file"
2023/03/06 08:19:51 DEBUG : old/file(0xc0005781c0): >openPending: err=<nil>
2023/03/06 08:19:51 DEBUG : old/file: >newRWFileHandle: err=<nil>
2023/03/06 08:19:51 DEBUG : old: Added virtual directory entry vAddFile: "file"
2023/03/06 08:19:51 DEBUG : old/file: >Open: fd=old/file (rw), err=<nil>
2023/03/06 08:19:51 DEBUG : old/: >Create: node=old/file, handle=&{old/file (rw)}, err=<nil>
2023/03/06 08:19:51 DEBUG : old/file: Attr: 
2023/03/06 08:19:51 DEBUG : old/file: >Attr: a=valid=1s ino=0 size=0 mode=-rw-r--r--, err=<nil>
2023/03/06 08:19:51 DEBUG : &{old/file (rw)}: Flush: 
2023/03/06 08:19:51 DEBUG : old/file(0xc0005781c0): RWFileHandle.Flush
2023/03/06 08:19:51 DEBUG : &{old/file (rw)}: >Flush: err=<nil>
2023/03/06 08:19:51 DEBUG : old/file: Setattr: a=Setattr [ID=0x16 Node=0x3 Uid=0 Gid=0 Pid=545184] atime=2023-03-06 08:19:51.226615647 +0100 CET atime=now mtime=2023-03-06 08:19:51.226615647 +0100 CET mtime=now handle=INVALID-0x0
2023/03/06 08:19:51 DEBUG : old/file: vfs cache: setting modification time to 2023-03-06 08:19:51.226615647 +0100 CET
2023/03/06 08:19:51 DEBUG : old/file: >Setattr: err=<nil>
2023/03/06 08:19:51 DEBUG : old/file: Attr: 
2023/03/06 08:19:51 DEBUG : old/file: >Attr: a=valid=1s ino=0 size=0 mode=-rw-r--r--, err=<nil>
2023/03/06 08:19:51 DEBUG : &{old/file (rw)}: Flush: 
2023/03/06 08:19:51 DEBUG : old/file(0xc0005781c0): RWFileHandle.Flush
2023/03/06 08:19:51 DEBUG : &{old/file (rw)}: >Flush: err=<nil>
2023/03/06 08:19:51 DEBUG : &{old/file (rw)}: Release: 
2023/03/06 08:19:51 DEBUG : old/file(0xc0005781c0): RWFileHandle.Release
2023/03/06 08:19:51 DEBUG : old/file(0xc0005781c0): close: 
2023/03/06 08:19:51 DEBUG : old/file: vfs cache: setting modification time to 2023-03-06 08:19:51.226615647 +0100 CET
2023/03/06 08:19:51 INFO  : old/file: vfs cache: queuing for upload in 1m0s
2023/03/06 08:19:51 DEBUG : old/file(0xc0005781c0): >close: err=<nil>
2023/03/06 08:19:51 DEBUG : &{old/file (rw)}: >Release: err=<nil>
2023/03/06 08:19:56 DEBUG : /: Attr: 
2023/03/06 08:19:56 DEBUG : /: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2023/03/06 08:19:56 DEBUG : /: ReadDirAll: 
2023/03/06 08:19:56 DEBUG : /: >ReadDirAll: item=5, err=<nil>
2023/03/06 08:19:56 DEBUG : /: Attr: 
2023/03/06 08:19:56 DEBUG : /: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2023/03/06 08:19:56 DEBUG : old/: Attr: 
2023/03/06 08:19:56 DEBUG : old/: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2023/03/06 08:19:59 DEBUG : /: Lookup: name="new"
2023/03/06 08:19:59 DEBUG : /: >Lookup: node=<nil>, err=no such file or directory
2023/03/06 08:19:59 DEBUG : /: Lookup: name="new"
2023/03/06 08:19:59 DEBUG : /: >Lookup: node=<nil>, err=no such file or directory
2023/03/06 08:19:59 DEBUG : /: Lookup: name="new"
2023/03/06 08:19:59 DEBUG : /: >Lookup: node=<nil>, err=no such file or directory
2023/03/06 08:19:59 DEBUG : /: Rename: oldName="old", newName="new", newDir=/
2023/03/06 08:19:59 DEBUG : old: Updating dir with new 0xc000b59ec0
2023/03/06 08:19:59 DEBUG : old: forgetting directory cache
2023/03/06 08:19:59 INFO  : old/file: vfs cache: renamed in cache to "new/file"
2023/03/06 08:19:59 INFO  : old: Removing directory
2023/03/06 08:19:59 INFO  : old: Removing directory
2023/03/06 08:19:59 INFO  : old/: vfs cache: renamed dir in cache to "new/"
2023/03/06 08:19:59 DEBUG : : Added virtual directory entry vDel: "old"
2023/03/06 08:19:59 DEBUG : : Added virtual directory entry vAddDir: "new"
2023/03/06 08:19:59 DEBUG : /: >Rename: err=<nil>
2023/03/06 08:19:59 DEBUG : /: Invalidating "new"
2023/03/06 08:20:04 DEBUG : /: Lookup: name="new"
2023/03/06 08:20:04 DEBUG : /: >Lookup: node=new/, err=<nil>
2023/03/06 08:20:04 DEBUG : new/: Attr: 
2023/03/06 08:20:04 DEBUG : new/: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2023/03/06 08:20:04 DEBUG : new/: ReadDirAll: 
2023/03/06 08:20:04 DEBUG : /: Attr: 
2023/03/06 08:20:04 DEBUG : /: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2023/03/06 08:20:04 DEBUG : new/: >ReadDirAll: item=3, err=<nil>
2023/03/06 08:20:04 DEBUG : new/: Attr: 
2023/03/06 08:20:04 DEBUG : new/: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2023/03/06 08:20:04 DEBUG : new/: ReadDirAll: 
2023/03/06 08:20:04 DEBUG : new/: >ReadDirAll: item=3, err=<nil>
2023/03/06 08:20:04 DEBUG : new/: Lookup: name="file"
2023/03/06 08:20:04 DEBUG : new/: >Lookup: node=old/file, err=<nil>
2023/03/06 08:20:04 DEBUG : old/file: Attr: 
2023/03/06 08:20:04 DEBUG : old/file: >Attr: a=valid=1s ino=0 size=0 mode=-rw-r--r--, err=<nil>
2023/03/06 08:20:04 DEBUG : new/: Remove: name="file"
2023/03/06 08:20:04 DEBUG : old/file: Remove: 
2023/03/06 08:20:04 DEBUG : new: Added virtual directory entry vDel: "file"
2023/03/06 08:20:04 DEBUG : old/file: >Remove: err=<nil>
2023/03/06 08:20:04 DEBUG : new/: >Remove: err=<nil>
2023/03/06 08:20:04 DEBUG : /: Remove: name="new"
2023/03/06 08:20:04 DEBUG : : Added virtual directory entry vDel: "new"
2023/03/06 08:20:04 DEBUG : /: >Remove: err=<nil>
2023/03/06 08:20:05 DEBUG : : Statfs: 
2023/03/06 08:20:05 DEBUG : : >Statfs: stat={Blocks:2621440 Bfree:2481820 Bavail:2481820 Files:1000000000 Ffree:1000000000 Bsize:4096 Namelen:255 Frsize:4096}, err=<nil>
2023/03/06 08:20:28 DEBUG : /: Attr: 
2023/03/06 08:20:28 DEBUG : /: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2023/03/06 08:20:42 DEBUG : vfs cache RemoveNotInUse (maxAge=120000000000, emptyOnly=false): item old/file not removed, freed 0 bytes
2023/03/06 08:20:42 DEBUG : vfs cache RemoveNotInUse (maxAge=120000000000, emptyOnly=false): item new/file not removed, freed 0 bytes
2023/03/06 08:20:42 INFO  : vfs cache: cleaned: objects 2 (was 2) in use 1, to upload 1, uploading 0, total size 0 (was 0)
2023/03/06 08:20:59 DEBUG : new/file: vfs cache: starting upload
2023/03/06 08:21:00 DEBUG : new/file: sha1 = da39a3ee5e6b4b0d3255bfef95601890afd80709 OK
2023/03/06 08:21:00 INFO  : new/file: Copied (new)
2023/03/06 08:21:00 DEBUG : new/file: vfs cache: fingerprint now "0,2023-03-06 07:19:51 +0000 UTC,da39a3ee5e6b4b0d3255bfef95601890afd80709"
2023/03/06 08:21:00 DEBUG : new/file: vfs cache: writeback object to VFS layer
2023/03/06 08:21:00 DEBUG : new/file: Not setting pending mod time 2023-03-06 08:19:51.226615647 +0100 CET as it is already set
2023/03/06 08:21:00 DEBUG : new: Added virtual directory entry vAddFile: "file"
2023/03/06 08:21:00 INFO  : new/file: vfs cache: upload succeeded try #1
2023/03/06 08:21:42 DEBUG : vfs cache RemoveNotInUse (maxAge=120000000000, emptyOnly=false): item old/file not removed, freed 0 bytes
2023/03/06 08:21:42 DEBUG : vfs cache RemoveNotInUse (maxAge=120000000000, emptyOnly=false): item new/file not removed, freed 0 bytes
2023/03/06 08:21:42 INFO  : vfs cache: cleaned: objects 2 (was 2) in use 0, to upload 0, uploading 0, total size 0 (was 0)
2023/03/06 08:22:42 INFO  : vfs cache RemoveNotInUse (maxAge=120000000000, emptyOnly=false): item old/file was removed, freed 0 bytes
2023/03/06 08:22:42 INFO  : new/file: vfs cache: removed cache file as Removing old cache file not in use
2023/03/06 08:22:42 DEBUG : new/file: vfs cache: removed metadata from cache as Removing old cache file not in use
2023/03/06 08:22:42 INFO  : vfs cache RemoveNotInUse (maxAge=120000000000, emptyOnly=false): item new/file was removed, freed 0 bytes
2023/03/06 08:22:42 INFO  : vfs cache: cleaned: objects 0 (was 2) in use 0, to upload 0, uploading 0, total size 0 (was 0)
2023/03/06 08:22:46 INFO  : Signal received: interrupt
2023/03/06 08:22:46 DEBUG : vfs cache: cleaner exiting
2023/03/06 08:22:46 ERROR : /mnt/tmp: Unmounted rclone mount
2023/03/06 08:22:46 INFO  : Exiting...
```

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
This turned out to be easier to fix that I was expecting...

Please give this a go

[v1.62.0-beta.6767.399716bce.fix-6809-vfs-rename](https://beta.rclone.org/branch/fix-6809-vfs-rename/v1.62.0-beta.6767.399716bce.fix-6809-vfs-rename/) on branch [fix-6809-vfs-rename](https://github.com/rclone/rclone/tree/fix-6809-vfs-rename) (uploaded in 15-30 mins)

**Comment by jdr85** [comment]
Thanks Nick, successfully tested with the fix-6809-vfs-rename and behavior is now as expected!

**Comment by ncw** [comment]
Thanks for testing.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.62
